### PR TITLE
Implementing Deepchem Model for Hamiltonian Neural networks

### PR DESCRIPTION
**Pull request by a-b-h-a-y-s-h-i-n-d-e**
## Description

This is the second PR followed by [first_PR](https://github.com/deepchem/deepchem/pull/4465)
this one adds the wrapper class, test_cases files and required changes to documentation.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ x ] Documentations (modification for documents)

## Checklist

- [ x ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ x ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ x ] Run `mypy -p deepchem` and check no errors
  - [ x ] Run `flake8 <modified file> --count` and check no errors
  - [ x ] Run `python -m doctest <modified file>` and check no errors
- [ x ] I have performed a self-review of my own code
- [ x ] I have commented my code, particularly in hard-to-understand areas
- [ x ] I have made corresponding changes to the documentation
- [ x ] I have added tests that prove my fix is effective or that my feature works
- [ x ] New unit tests pass locally with my changes
- [ x ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/hnn.py [hunk 1]
```diff
@@ -188,3 +190,82 @@ def symplectic_gradient(self, z: torch.Tensor) -> torch.Tensor:
         dp_dt = -dH_dq
 
         return torch.cat([dq_dt, dp_dt], dim=-1)
+
+
+class HNNModel(TorchModel):
+    """Hamiltonian Neural Network wrapper model which inherits TorchModel.
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/hnn.py:196]: Can you please explain more about this model and include references?

File: deepchem/models/torch_models/hnn.py [hunk 2]
```diff
@@ -188,3 +190,82 @@ def symplectic_gradient(self, z: torch.Tensor) -> torch.Tensor:
...
+                 d_hidden: Tuple[int, ...] = (32, 32),
+                 activation_fn: str = 'tanh',
+                 **kwargs) -> None:
+        """Initialize HNNModel."""
+        model = HNN(d_input=d_input,
+                    d_hidden=d_hidden,
+                    activation_fn=activation_fn)
+        super().__init__(model, loss=L2Loss(), **kwargs)
+
+    def predict_on_batch(self, X):
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/hnn.py]: add type hints

File: deepchem/models/torch_models/hnn.py [hunk 3]
```diff
@@ -188,3 +190,82 @@ def symplectic_gradient(self, z: torch.Tensor) -> torch.Tensor:
         dp_dt = -dH_dq
 
         return torch.cat([dq_dt, dp_dt], dim=-1)
+
+
+class HNNModel(TorchModel):
+    """Hamiltonian Neural Network wrapper model which inherits TorchModel.
+
+    Parameters
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/hnn.py:203]: Add usage examples

File: deepchem/models/torch_models/hnn.py [hunk 4]
```diff
@@ -188,3 +190,82 @@ def symplectic_gradient(self, z: torch.Tensor) -> torch.Tensor:
...
+        """
+        self.model.eval()
+        X_tensor = torch.tensor(X,
+                                dtype=torch.float32,
+                                device=self.device,
+                                requires_grad=False)
+        predictions = self.model.symplectic_gradient(X_tensor)
+        return predictions.detach().cpu().numpy()
+
+    def predict_hamiltonian(self, X):
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/hnn.py]: type hints

File: deepchem/models/torch_models/hnn.py (hunk 1 above)
Comment (a-b-h-a-y-s-h-i-n-d-e) [reply deepchem/models/torch_models/hnn.py:196]: done

File: deepchem/models/torch_models/hnn.py (hunk 2 above)
Comment (a-b-h-a-y-s-h-i-n-d-e) [reply deepchem/models/torch_models/hnn.py]: done

File: deepchem/models/torch_models/hnn.py (hunk 4 above)
Comment (a-b-h-a-y-s-h-i-n-d-e) [reply deepchem/models/torch_models/hnn.py]: done

File: deepchem/models/torch_models/hnn.py (hunk 3 above)
Comment (a-b-h-a-y-s-h-i-n-d-e) [reply deepchem/models/torch_models/hnn.py:203]: done

File: deepchem/models/torch_models/hnn.py [hunk 5]
```diff
@@ -188,3 +193,113 @@ def symplectic_gradient(self, z: torch.Tensor) -> torch.Tensor:
...
+        np.ndarray
+            Hamiltonian values of shape (n_samples,).
+        """
+        self.model.eval()
+        X_tensor = torch.tensor(X, dtype=torch.float32, device=self.device)
+        H = self.model.hamiltonian(X_tensor)
+        return H.detach().cpu().numpy()
+
+    def symplectic_gradient(self, z: torch.Tensor) -> torch.Tensor:
+        """Compute symplectic gradient using Hamilton's equations.
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/hnn.py:265]: Can you please add more about symplectic gradients in docstrings? I think adding equations and usage examples would add a lot.

File: deepchem/models/torch_models/hnn.py [hunk 6]
```diff
@@ -188,3 +193,113 @@ def symplectic_gradient(self, z: torch.Tensor) -> torch.Tensor:
...
+                 d_hidden: Tuple[int, ...] = (32, 32),
+                 activation_fn: str = 'tanh',
+                 **kwargs) -> None:
+        """Initialize HNNModel."""
+        model = HNN(d_input=d_input,
+                    d_hidden=d_hidden,
+                    activation_fn=activation_fn)
+        super().__init__(model, loss=L2Loss(), **kwargs)
+
+    def predict_on_batch(
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/hnn.py]: This model is not fully compatible with Deepchem's API. After training, you should be able to call model.evaluate, model.predict, etc., and expect some results (preds, loss). I think you are close, but it needs a bit of work for complete integration.

File: deepchem/models/torch_models/tests/test_hnn.py [hunk 7]
```diff
@@ -28,3 +31,91 @@ def test_symplectic_gradient_shape():
...
+
+
+@pytest.mark.torch
+@pytest.mark.skipif(not has_torch, reason="PyTorch is not installed")
+def test_hnnmodel_train_and_predict():
+    """Test HNNModel can fit to simple synthetic data and predict correct shape."""
+    from deepchem.models.torch_models import HNNModel
+
+    # small synthetic data (q, p) and their derivatives
+    x = np.random.randn(10, 2).astype(np.float32)
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/tests/test_hnn.py:43]: It would be nice to add some tests with real-world physics simulations, in addition to synthetic data.

File: deepchem/models/torch_models/hnn.py (hunk 6 above)
Comment (a-b-h-a-y-s-h-i-n-d-e) [reply deepchem/models/torch_models/hnn.py:248]: Thanks for the suggestions.

the .predict method is working ( I re-checked that ) , it accepts dataset object and then does the forward pass on model and returns the results
I have not checked the .evaluate method, so will check that now

Comment (a-b-h-a-y-s-h-i-n-d-e) [reply deepchem/models/torch_models/hnn.py]: Thanks for suggestions, looks like one change in base class method ( forward pass ) is fixing this all, so I have made those changes now

File: deepchem/models/torch_models/tests/test_hnn.py (hunk 7 above)
Comment (a-b-h-a-y-s-h-i-n-d-e) [reply deepchem/models/torch_models/tests/test_hnn.py:43]: great suggestion
I think, simulating a whole experiment will take more LOC, the author of paper has provided scripts to simulate it which generally needs 3 functions, but lets see how we can do this in some other ways

File: deepchem/models/torch_models/hnn.py (hunk 5 above)
Comment (a-b-h-a-y-s-h-i-n-d-e) [reply deepchem/models/torch_models/hnn.py:265]: done

File: deepchem/models/torch_models/hnn.py [hunk 8]
```diff
@@ -138,6 +138,8 @@ def symplectic_gradient(self, z: torch.Tensor) -> torch.Tensor:
 
         The gradients are computed using automatic differentiation to ensure
         exact computation of the partial derivatives.
+        dq/dt = ∂H/∂p
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/hnn.py:141]: Will the `∂` symbol render correctly in docs?

File: deepchem/models/torch_models/tests/test_hnn.py [hunk 9]
```diff
@@ -27,4 +30,92 @@ def test_symplectic_gradient_shape():
...
+                                          mode='regression')
+
+    model = HNNModel(batch_size=5, learning_rate=1e-3)
+
+    model.fit(dataset, nb_epoch=1000)
+    pred = model.predict_on_batch(x)
+
+    score = regression_metric.compute_metric(dx, pred)
+
+    assert score < 0.06, "HNNModel failed to overfit small dataset"
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/tests/test_hnn.py:121]: It would be nice to add a test for a physics related system.

**Comment by JoseAntonioSiguenza** [comment]
Minor comments on docstrings and a test case. CI looks good. LGTM.

File: deepchem/models/torch_models/hnn.py (hunk 8 above)
Comment (a-b-h-a-y-s-h-i-n-d-e) [reply deepchem/models/torch_models/hnn.py:141]: I am not aware of it actually, but then in general what will be alternate symbol for partial derivatives?

File: deepchem/models/torch_models/tests/test_hnn.py (hunk 9 above)
Comment (a-b-h-a-y-s-h-i-n-d-e) [reply deepchem/models/torch_models/tests/test_hnn.py:121]: Hmm, I am working on it will create new PR for that test case

File: deepchem/models/torch_models/hnn.py (hunk 8 above)
Comment (a-b-h-a-y-s-h-i-n-d-e) [reply deepchem/models/torch_models/hnn.py:141]: `{\partial{H}}/{\partial{p}}` This is the latex representation for that equation, but I think the symbol `∂` should render, let's see what @rbharath says.
